client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentEventReaderImpl.java

```diff
@@ -0,0 +1,380 @@
+        log.trace("SetEndOffset {}", offset);
+        Preconditions.checkArgument(offset >= 0);
+        Exceptions.checkNotClosed(asyncInput.isClosed(), this);
+        this.endOffset = offset;
+    }
+
+    @Override
+    @Synchronized
+    public long getEndOffset() {
+        return endOffset;
+    }
+
+    @Override
+    @Synchronized
+    public CompletableFuture<ByteBuffer> readAsync() throws EndOfSegmentException {
+
+        // prepare a promise
+        if (cursorOffset >= this.endOffset) {
+            log.debug("All events up to the configured end offset:{} have been read", endOffset);
+            throw new EndOfSegmentException(END_OFFSET_REACHED);
+        }
+        CompletableFuture<ByteBuffer> promise = readState.reset(cursorOffset);
+        promise.whenComplete(this::readCompleted);
+
+        // kick off or continue processing the request
+        issueRequestIfNeeded();
+
+        return promise;
+    }
+
+    @Synchronized
+    private void readCompleted(ByteBuffer result, Throwable th) {
+        if (th != null) {
+            // an event could not be read; don't update the cursor position.
+            log.trace("Failed to read an event @ " + cursorOffset, th);
+
+            // cancel any outstanding request now to ensure that handleRequest doesn't affect the read state
+            if (outstandingRequest != null && !outstandingRequest.isDone()) {
+                log.trace("Cancel outstanding read request for segment {}", asyncInput.getSegmentId());
+                outstandingRequest.cancel(false);
+                outstandingRequest = null;
+            }
+        }
+        if (result != null) {
+            // an event was read successfully; update the cursor position.
+            cursorOffset = readState.offset;
+            log.trace("Read an event @ {}", cursorOffset);
+        }
+    }
+
+    @VisibleForTesting
+    ReadState getReadState() {
+        return readState;
+    }
+
+    static class ReadState {
+        private final ByteBuffer headerReadingBuffer = ByteBuffer.allocate(WireCommands.TYPE_PLUS_LENGTH_SIZE);
+        private CompletableFuture<ByteBuffer> outstandingPromise = CompletableFuture.completedFuture(null);
+        private ByteBuffer resultBuffer;
+        private long offset;
+
+        /**
+         * Gets the next offset for reading purposes.
+         */
+        public long getOffset() {
+            return offset;
+        }
+
+        /**
+         * Resets the read state in preparation for a new read.
+         * @param offset the offset to prepare for.
+         */
+        public CompletableFuture<ByteBuffer> reset(long offset) {
+            if (!this.outstandingPromise.isDone()) {
+                this.outstandingPromise.cancel(false);
+            }
+            this.outstandingPromise = new CompletableFuture<>();
+            this.headerReadingBuffer.clear();
+            this.resultBuffer = null;
+            this.offset = offset;
+            return this.outstandingPromise;
```

Comment:
This is vending internal state to anyone in the same package. The correctness of this cannot be validated. 
Perhaps this should be private?